sdk/core/azure-core/inc/azure/core/http/http.hpp

```diff
@@ -135,154 +155,208 @@ namespace Azure { namespace Core { namespace Http {
     Stream,
   };
 
-  // parses full url into protocol, host, port, path and query.
+  // Url represent the location where a request will be performed. It can be parsed and init from an
+  // a string that contains all Url parts (scheme, host, path, etc).
   // Authority is not currently supported.
-  class URL {
+  class Url {
+    // Let Request class to be able to set retry enabled ON
+    friend class Request;
+
   private:
     std::string m_scheme;
     std::string m_host;
-    std::string m_port;
+    int m_port{-1};
+    // m_path is encoded
     std::string m_path;
+    std::string m_fragment;
+    // query parameters are all decoded
     std::map<std::string, std::string> m_queryParameters;
+    std::map<std::string, std::string> m_retryQueryParameters;
+    bool m_retryModeEnabled{false};
+
+    /*********  private static methods for all instances *******/
+    static std::string Decode(const std::string& value);
+    static std::string EncodeHost(const std::string& host);
+    static std::string EncodePath(const std::string& path);
+    static std::string EncodeQuery(const std::string& query);
+    static std::string EncodeFragment(const std::string& fragment);
+    static std::string EncodeImpl(
+        const std::string& source,
+        const std::function<bool(int)>& should_encode);
+
+    void StartRetry()
```

Comment:
The Url should not know about the Retry headers.  This makes the Url not shareable.

Reply:
Url handles retry query parameters.
Request handles retry headers.

Request.StartRetry() calls Url.StartRetry().

I don't want to have Request.AddQueryParameter() and Url.AddQueryParameter(), so all query parameters logic was moved to be in Url

Reply:
Why not just remove retry headers and retry query parameters? We can remove/add/overwrite headers and query parameters anytime we want. I really cannot think of anything that cannot be done without retry headers or retry query parameter.s

Reply:
I think one of the reasons why we need retry mode is for logging policy.
When a request is received by the Pipeline, we don't want to mutate the data in it.  We can add more staff to it (like query/headers) but we don't want to override values from the request.
This is because when logging the request, we want to show the values as they were received.

There might be more features that requires the Retry mode feature from the Request.  It's just a way we can override values temporally.

Reply:
I mentioned this offline, but will re-iterate it. Merging the request/retry behavior into the Url class goes against single responsibility which impacts the implementation in a couple of less-than-ideal ways.

Creating a "snapshot" to revert changes to the URL should be done by the caller, either by creating a copy of the Url, or possibly getting the url as string, or with some other way.

That said, I think it is OK to defer this decision until later to unblock this PR, but we should strongly consider fixing it at some point (file an issue?).

Reply:
> Merging the request/retry behavior into the Url class goes against single responsibility which is also impacting the implementation in a couple of less-than-ideal ways.

> Needing to create a "snapshot" to revert changes to the URL (as part of retry) should be done by the caller, either by creating a copy of the Url, or possibly getting the url as string, or with some other way.

I couldn't agree more.

Reply:
https://github.com/Azure/azure-sdk-for-cpp/issues/580